eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -357,7 +393,28 @@ instance
   ) =>
   ShelleyEraImp (ShelleyEra c)
   where
-  emptyImpNES = emptyShelleyImpNES
+  initImpNES = initShelleyImpNES
+
+  impSatisfyNativeScript script = do
+    keyPairs <- gets impKeyPairs
+    let
+      satisfyMOf m []
+        | m <= 0 = Just mempty
+        | otherwise = Nothing
+      satisfyMOf m (x : xs) =
+        case satisfyScript x of
+          Nothing -> satisfyMOf m xs
+          Just kps -> do
+            kps' <- satisfyMOf (m - 1) xs
+            Just $ kps <> kps'
+      satisfyScript = \case
+        RequireSignature keyHash -> do
+          keyPair <- Map.lookup keyHash keyPairs
+          Just $ Map.singleton keyHash keyPair
+        RequireAllOf ss -> satisfyMOf (length ss) ss
+        RequireAnyOf ss -> satisfyMOf 1 ss
+        RequireMOf m ss -> satisfyMOf m ss
+    pure $ satisfyScript script
```

Comment:
Could we reuse some of this in `impAllegraSatisfyNativeScript`?

Reply:
Not at this point. I have plans for improving how we deal with native scripts which would allow us to reuse this logic. That will come in the future, when we have a bit more time. However, at this point, `MultiSigs` and `Timelocks` are two totally separate types that just happen to have pattern synonyms with same names.